### ISSUE TITLE
[UX] Bigger up/down arrows on number fields.

### ISSUE THREAD

**Issue by klonos** [backdrop/backdrop-issues#2685]
### Steps to reproduce

These are a couple of examples of how our number fields are rendered (this is on Firefox, on Windows):

![image](https://cloud.githubusercontent.com/assets/2423362/25991442/8165e3c4-3746-11e7-85b2-76a89ee60ce2.png)

![image](https://cloud.githubusercontent.com/assets/2423362/25991468/913a68d8-3746-11e7-8654-23e6da5f38c7.png)

### Actual behavior

The up/down arrows are really small and they do not match the rest of the admin UI style. They are also very small target for touch devices.

### Expected behavior

Bigger, finger-friendly up/down buttons.

**Comment by klonos** [comment]
I like [this](https://jqueryui.com/spinner):

![image](https://cloud.githubusercontent.com/assets/2423362/25991745/bd650caa-3747-11e7-9d9b-326ee80c7b4b.png)

...but it bothers me that it doesn't use `type="number"` ...or can it?

**Comment by klonos** [comment]
Related: https://css-tricks.com/numeric-inputs-a-comparison-of-browser-defaults/

**Comment by klonos** [comment]
Something like [this](https://css-tricks.com/number-increment-buttons) perhaps?:

![buttonincrememters](https://cloud.githubusercontent.com/assets/2423362/25992142/ff04ad90-3748-11e7-937d-605e0792ddcd.png)

...but the buttons could be styled to match Seven and perhaps placed to the left/right of the number field.

**Comment by Dinornis** [comment]
Thanks for bringing this up.  Yes the default browser arrows are too small on desktops.
I like the Spinner up/down buttons.

**Comment by quicksketch** [comment]
Here's a pretty good article covering using jQuery UI's Spinner on both text and number input elements: https://www.tjvantoll.com/2012/07/15/native-html5-number-picker-vs-jquery-uis-spinner-which-to-use/

It's worth noting that jQuery UI Spinner is already in core, so if we want to use it, it's a matter of integrating it properly.

**Comment by olafgrabienski** [comment]
> The up/down arrows are (...) also very small target for touch devices.

For `input type="number"`, most touch devices don't display the arrows but a number pad, which is fine in my opinion. Reading the article linked by @quicksketch, jQuery Spinner in contrast seems not to show the number pad.

**Comment by jenlampton** [comment]
We need to be careful about styling away the default browser style. It's possible to really hurt the user experience by doing so. Having a number field look the same on all websites everywhere means that when someone sees one, they know how to use it. If we make Backdrop's number fields look different from everything else, users will not recognize it instantly.

I'd be in favor of adopting something else that's wildly popular (like maybe jQuery UI's Spinner) as long as we can confirm it is an improvement across all browsers, and only if we can do so in a way that won't break mobile devices opening the keypad.

**Comment by klonos** [comment]
I agree that using a 3rd party library and taking advantage of PIE is better in general, but in this case I would like it better if we could either find a solution or customize one that gives more touch surface for the +/- buttons. I mean Spinner is definitely an improvement over the current browser defaults and we should use that if we want something implemented quickly. But if we managed to somehow get those buttons on either side of the field, we would make it easier for those that use touch devices.

**Comment by jenlampton** [comment]
As stated before, touch surface is a non issue because touch devices open a
number keyboard instead.

On Jun 9, 2017 3:01 AM, "Gregory Netsas" <notifications@github.com> wrote:

I agree that using a 3rd party library and taking advantage of PIE is
better in general, but in this case I would like it better if we could
either find a solution or customize one that gives more touch surface for
the +/- buttons. I mean Spinner is definitely an improvement over the
current browser defaults and we should use that if we want something
implemented quickly. But if we managed to somehow get those buttons on
either side of the field, we would make it easier for those that use touch
devices.

—
You are receiving this because you commented.
Reply to this email directly, view it on GitHub
<https://github.com/backdrop/backdrop-issues/issues/2685#issuecomment-307349210>,
or mute the thread

**Comment by jlfranklin** [comment]
Is it possible for a contrib module to override the core theming for controls?  Presenting the browser default for these things, but allowing contrib to change it would be a good compromise.

**Comment by klonos** [comment]
>As stated before, touch surface is a non issue because touch devices open a number keyboard instead.

I just tested on a touch laptop and the num keyboard does not pop up.

**Comment by jenlampton** [comment]
I expect, just like with select lists, it will be nearly impossible to
style these elements at all, and for good reason: for consistency across
the whole internet.

There is currently no core theming on number fields. What you are seeing
right now is the behavior of whatever browser you are using (backdrop has
no say in the matter). If your browser doesn't handle them in a way that is
sensible (@klonos) it might be worthwhile creating an issue with the
browser too, so that everyone else can benefit from a fix :)

That said, it is possible to use JavaScript to replace these standard
elements with something else, and then style the new element to do
something unique to backdrop.

(Keep in mind that if we bury the standard behavior we then won't benefit
from other improvements browsers will implement over time.)

Personally, I'd prefer to stick with some standard, maybe something like
jQuery UI. However, it looks like jQuery UI currently breaks the number
keypad behavior on mobile devices, and since that's why we put in the
number fields in the first place, it sounds like a non-starter.

Are there other examples of systems that have encountered this issue? What
is the rest of internet doing about it?

**Comment by jlfranklin** [comment]
> That said, it is possible to use JavaScript to replace these standard elements with something else, and then style the new element to do something unique to backdrop. 

This is closer to what I expect a contrib module to provide.  Similar to how [Better Select](https://www.drupal.org/project/betterselect) replaces a select control with checkboxes or radios, a contrib module could add `$element['attached']['js'][] = 'myspinnylib.js'` in a hook, or replace the entire element with different HTML that includes the input field plus two buttons or something.

Backdrop's core implementation of a stock HTML5 number field is, IMHO, best for the default.

**Comment by klonos** [comment]
> Are there other examples of systems that have encountered this issue?

Well, I don't know about other browsers, but in firefox that issue was reported 4 years ago 😞 https://bugzilla.mozilla.org/show_bug.cgi?id=947365

> What is the rest of internet doing about it?

That's what I've found:

https://developer.apple.com/ios/human-interface-guidelines/ui-controls/steppers

![stepper](https://user-images.githubusercontent.com/2423362/26990530-39a1c8f4-4d9a-11e7-8191-88254ee3b634.png)

http://www.virtuosoft.eu/code/bootstrap-touchspin
https://github.com/istvan-ujjmeszaros/bootstrap-touchspin

![image](https://user-images.githubusercontent.com/2423362/26990733-12a1b056-4d9b-11e7-83cc-9b03acd4cc1c.png)

**Comment by jenlampton** [comment]
> This is closer to what I expect a contrib module to provide. [...] Using a stock HTML5 number field is, IMHO, best for core's default implementation.
 
This is how I feel too, at least until there's a clear winner in terms of suitable alternatives. Today there's nothing to stop a contrib module from adding TouchSpin, for example.
